### PR TITLE
Add version information to API

### LINKED ISSUE #47: Add version information to API
C library provides version info: https://roc-streaming.org/toolkit/docs/api/reference.html#roc-version

It would be nice to add this feature to the go library too. We need to provide two versions actually: version of the native library (libroc), and version of the bindings (which may be slightly different).

Versions of the bindings can be obtained using https://pkg.go.dev/runtime/debug

### PR DESCRIPTION
**Pull request by dhanusaputra**
close #47

### CODE REVIEW

File: roc/version.go
Comment (gavv) [inline roc/version.go:20]: nit:

* we can remove pointer
* librock => libroc :)

File: roc/version.go
Comment (gavv) [inline roc/version.go:21]: * Can we parse string into components and use SemanticVersion for this field too?
* Binding => Go bindings

File: roc/version.go
Comment (gavv) [inline roc/version.go:34]: nit:

* we can remove the pointer to make it clear that it's never nil
* maybe rename it from GetVersion to just Version for simplicity? roc.Version() looks better IMHO
* maybe we can also rename struct Version to struct Versions to avoid confusion with the function

File: roc/version_test.go
Comment (gavv) [inline roc/version_test.go:12]: We can also test that:

* at least one of the version components is non-zero
* major versions of C and Go are the same
* minor version of Go >= minor version of C

These invariants should be always true.

**Review by gavv — COMMENTED** [review]
Thanks for PR! A few minor comments

File: roc/version_test.go
Comment (dhanusaputra) [reply roc/version_test.go:12]: for Go version seems not able to get it during unit testing
only available after it's built
so don't check it

Comment (dhanusaputra) [reply roc/version_test.go]: btw not sure how to build this, there is no main.go

**Comment by gavv** [comment]
BTW welcome to matrix chat! https://roc-streaming.org/toolkit/docs/about_project/contacts.html

File: roc/version.go
Comment (dhanusaputra) [reply roc/version.go]: instead of notify us via CI, maybe can try push the newer version

Comment (dhanusaputra) [reply roc/version.go]: seems need to give permission to github actions first before it can push some changes
so updated to check only

File: roc/version.go
Comment (Asalle) [inline roc/version.go:40]: I think it's generally beneficial to check what do functions return, here for example some weird stuff can happen and C.roc_version_get can fail, then we'll just panic with no explanation. A more graceful shutdown stating that roc_version_get crapped out would be preferable.

**Comment by Asalle** [comment]
Amazing job! It would make it even better if you included some motivation / `closes #XX` in the commit message to document it a bit more.

File: roc/version_test.go
Comment (Asalle) [inline roc/version_test.go:9]: Great test, although it does only test a successful path. I can see that the way `Version` func is implemented you cannot really inject any test strings into it and test failures here. Have you thought about decoupling it to make it possible? There are a lot of moving parts in `Version` and it can crap out in e. g. lines 40, 47, 51, 52 and 53. I am not sure if it is vital, but I would suggest it.

File: roc/version_test.go
Comment (Asalle) [reply roc/version_test.go]: weird, the `roc_get_version` should be definitely available during unit tests, just to smoke test, do you have libroc installed? I wrote a simple sanity test to check it:
```
package main

/*
#cgo LDFLAGS: -lroc

#include <roc/version.h>
*/
import "C"

import (
        "fmt"
)

func main() {
        var cVersion C.struct_roc_version
        C.roc_get_version(&cVersion)
        fmt.Print(cVersion)
}
```
My output is 
```
 go run /tmp/test.go
{0 2 1}%    
```

I think I have an old version of roc where the method for getting the version is named differently, please rename it if it does not work. Let me know if you have any problems with this!

Comment (Asalle) [reply roc/version_test.go]: > btw not sure how to build this, there is no main.go

there's no main.go because it's a library, basically just a set of functions that can be used by applications that do have main.go. Unit tests pull on those methods (by having a func main declared somewhere under the hood), so if you want to test something, write a test for it with go unit testing framework or make a small program like I did, but don't commit it unless it's in the unit test form.

Comment (dhanusaputra) [reply roc/version_test.go]: > I think I have an old version of roc where the method for getting the version is named differently, please rename it if it does not work. Let me know if you have any problems with this!

yes in newer version it's renamed https://roc-streaming.org/toolkit/docs/development/changelog.html#version-0-2-2-feb-27-2023

> there's no main.go because it's a library, basically just a set of functions that can be used by applications that do have main.go. Unit tests pull on those methods (by having a func main declared somewhere under the hood), so if you want to test something, write a test for it with go unit testing framework or make a small program like I did, but don't commit it unless it's in the unit test form.

got it, thanks for your explanation!

File: roc/version.go
Comment (dhanusaputra) [reply roc/version.go:40]: yup agree but seems from [API doc](https://roc-streaming.org/toolkit/docs/api/reference.html#roc-version) the roc_version_get will not return error, please CMIIW thanks

File: roc/version_test.go
Comment (dhanusaputra) [reply roc/version_test.go:9]: > Have you thought about decoupling it to make it possible? 

yes already thought about this, but seems ok to me to ignore and try to return as much as possible instead of returning error

> There are a lot of moving parts in Version and it can crap out in e. g. lines 40, 47, 51, 52 and 53. I am not sure if it is vital, but I would suggest it.

yes those lines can fail when parsing invalid `bindingsVersion`
let me try add 1 unit test to check whether the `bindingsVersion` is valid or not so we can check it early in unit test

File: roc/version.go
Comment (Asalle) [reply roc/version.go:40]: I see, then I guess the API designers agreed on some null-like version (e. g. 0.0.0) that will indicate an error? @gavv would appreciate your input

File: roc/version_test.go
Comment (Asalle) [reply roc/version_test.go:11]: > yes already thought about this, but seems ok to me to ignore and try to return as much as possible instead of returning error

In general, I would not think it's ok to do it, but I can see your motivation, because `Version` returns {0,0,0} in case of error anyway. Let's see how that discussion plays out, so depending on what we decide, you would leave this func as is or rewrite it a bit to decouple getting the version from C and parsing it to allow for more modularity and testability.

File: roc/version_test.go
Comment (Asalle) [inline roc/version_test.go:20]: Thanks for the update! Does it pass locally for you? I'm not sure where does the var `bindingsVersion` come from, and on the CI it this test seems to be skipped. I haven't run it locally, let me try it later.

File: roc/version_test.go
Comment (Asalle) [reply roc/version_test.go]: did that small test of mine work for you? if not - let me know, I'd be happy to help :rocket:

Comment (dhanusaputra) [reply roc/version_test.go]: yes the test works

File: roc/version_test.go
Comment (dhanusaputra) [reply roc/version_test.go:20]: > Does it pass locally for you?

yes

> I'm not sure where does the var bindingsVersion come from, and on the CI it this test seems to be skipped

it's from [latest release](https://github.com/roc-streaming/roc-go/releases), it's skipped because need to push changes with tag

File: roc/version_test.go
Comment (gavv) [reply roc/version_test.go]: @Asalle 

> weird, the roc_get_version should be definitely available during unit tests

I think what was missing is **bindings** version (which we tried to get using ReadBuildInfo), not libroc version. See https://github.com/roc-streaming/roc-go/pull/61#discussion_r1145008792

File: roc/version.go
Comment (gavv) [reply roc/version.go]: > instead of notify us via CI, maybe can try push the newer version

What do you mean by pushing new version?

File: roc/version.go
Comment (gavv) [reply roc/version.go:40]: @Asalle it just can't fail, because it returns version hard-coded during build; thus no error code

File: roc/version_test.go
Comment (gavv) [reply roc/version_test.go:11]: >  let me try add 1 unit test to check whether the bindingsVersion is valid or not so we can check it early in unit test

I like this idea. User interface is simpler (no need to check for errors from roc.Version), and at the same time our test suite guarantees that failures are not possible (by validating hard-coded version).

The only downside is that the test duplicates parsing logic. If we update version.go and forget version_test.go, then the test's validation may become irrelevant.

So I'd suggest to refactor it to reuse parsing code. One way to do it:

1. extract private function parseVersion that accepts string, returns version triple, or panics on invalid string format
2. in roc.Version, use parseVersion
3. in tests, check that:
  - parseVersion works on valid input
  - parseVersion panics on invalid input
  - parseVersion works in hard-coded bindingsVersion

File: roc/version.go
Comment (gavv) [inline roc/version.go:23]: I think it will be more clear if we rename "Library" to "Native"

File: roc/version_test.go
Comment (gavv) [inline roc/version_test.go:17]: nit: I suggest to rename tests to something like this:

* TestVersion => TestVersion_Get
* Test_parseVersion => TestVersion_Parse

For consistency with other tests, where we name them `Test<File>_<Component>`.

File: roc/version_test.go
Comment (gavv) [inline roc/version_test.go:83]: nit: we can simplify this a bit:

```go
			if tt.wantPanic {
				require.Panics(t, func() { parseVersion(tt.args.s) })
			} else {
				require.Equal(t, tt.want, parseVersion(tt.args.s))
			}
```

(we can require that either wantPanic is true of want is set; we don't need NotPanics I think)

File: roc/version_test.go
Comment (gavv) [reply roc/version_test.go:9]: +1 from me, @Asalle feel free to resolve this

**Comment by gavv** [comment]
Added 4 minor comments, otherwise looks fine. Waiting for review from Asa.

**Review by gavv — APPROVED** [review]
LGTM

File: roc/version_test.go
Comment (Asalle) [reply roc/version_test.go]: Oh I see, had to get some context to review this PR properly

File: roc/version.go
Comment (Asalle) [reply roc/version.go:40]: oh, I see, got it, then the implementation does makes sense. Good job @dhanusaputra!

File: roc/version_test.go
Comment (Asalle) [reply roc/version_test.go]: I remember injecting an env var during build [like so, with ldflags](https://stackoverflow.com/questions/68051479/how-to-go-build-ldflags-with-custom-environment-variable), didn't try it with libs though, should work with go build, right? Worth a try! Seems like [you can inject that stuff into packages as well, not just main](https://goenning.net/2017/01/25/adding-custom-data-go-binaries-compile-time/).

**Review by Asalle — APPROVED** [review]
Again, amazing job and thanks for your contribution!

File: roc/version_test.go
Comment (gavv) [reply roc/version_test.go]: Discussion moved to chat

**Pull request merged**
